gnu/man/sum.man
NAME
     sum - checksum and count the blocks in a file

SYNOPSIS
     sum [-s] [--sysv] [file...]

DESCRIPTION
     This manual page documents the GNU version of sum.  sum computes a 16-bit
     checksum  for each named file, or the standard input if none are given or
     when a file named `-' is given.  It prints the  checksum  for  each  file
     along  with  the  number of blocks in the file (rounded up), and the file
     name unless no arguments were given.  By default, the  GNU  sum  computes
     checksums  using  an  algorithm  that  is compatible with the BSD sum and
     prints file sizes in units of 1K blocks.

     OPTIONS

     -s, --sysv
          Compute checksums using an algorithm that is compatible with the one
          the  System  V  sum uses by default and print file sizes in units of
          512-byte blocks instead of 1K.

     The long-named options can be introduced with `+' as well  as  `--',  for
     compatibility with previous releases.  Eventually support for `+' will be
     removed, because it is incompatible with the POSIX.2 standard.


gnu/ChangeLog
Sun Feb  2 21:22:01 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* Version 1.2.

	* nl.c: Support multiple files and "-" for stdin.
	(main): Check for read and write errors.
	(nl_file): New function.

Wed Jan 29 10:09:10 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c (main): -t option was called -b in getopt spec.
	(validate): Don't warn that set1 is longer than set2.

	* tr.c: Rename --sysv-string2-truncate to --truncate-string1.

Fri Jan 17 16:29:05 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* nl.c: New program from bin-src.

	* nl.c (main): Use a struct linebuffer for line_buf.
	(process_file): Use readline instead of fgets, to preserve NULs.
	(check_section): Use memcmp instead of strncmp.
	(proc_text): Print line_buf with fwrite instead of printf.

	* nl.c (main): Usage message if too many args given.  Check
	for error in closing input file.  Lengths of section delimiter
	strings were 1 too large.  Take separator_str into account in
	length of print_no_line_fmt.
	(build_print_fmt): Allocate space for print_fmt, in case
	separator_str is long.
	(proc_text): A blank line is one that contains nothing, not
	even nonprinting characters.

Fri Jan 17 01:04:22 1992  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* All programs: Document `--' instead of `+' to introduce
	long-named options, in usage messages.

	* sum.c (bsd_sum_file): Renamed from sum_file.
	(sysv_sum_file): New function.
	(main): Recognize an option to select between the algorithms.

Sun Jan  5 17:41:18 1992  Jim Meyering (meyering at churchy.gnu.ai.mit.edu)

	* pr.c (close_file, print_page): Fixed bug that had the command
	yes |head |pr -t printing "yyyyyyyyyy".
	* (print_page): Fixed bug that had pr -3 -a printing two too few
	trailer lines per page.
	* (main): Added restriction that -a and -m are incompatible.
	Although the POSIX spec doesn't explicitly say they shouldn't
	be used together, it says -a modifies the -column option and
	that -column shouldn't be used with -m.

Thu Jan  2 15:23:59 1992  David J. MacKenzie  (djm at albert.gnu.ai.mit.edu)

	* nl.c: Include regex.h after, not before, sys/types.h.

Thu Jan  2 12:18:10 1992  Tom Lord  (lord at geech.gnu.ai.mit.edu)

	* sort.c (fillbuf) return bytes buffered instead of bytes read.

Fri Dec 27 22:53:36 1991  Jim Kingdon  (kingdon at geech.gnu.ai.mit.edu)

	* sort.c (LINEALLOC): New #define.
	(struct lines): New field ``limit''.
	(initlines): Set it from new arg ``limit''.
	(sort, mergefps, checkfp): Pass new arg to initlines().
	(findlines): Don't realloc past lines->limit.

Tue Dec 24 01:24:03 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c, system.h: Change POSIX ifdefs to
	HAVE_UNISTD_H and _POSIX_VERSION.

	* xwrite.c: Change POSIX ifdef to HAVE_UNISTD_H.

Sat 14 Dec 1991 11:46:42  Jim Meyering  (meyering at wombat)

	* tr.c: Fixed an inaccurate comment on posix_pedantic.

Thu 12 Dec 1991 21:15:20  Jim Meyering  (meyering at hal)

	* tr.c: Changed underscores to hyphens in long option name
	"sysv_string2_truncate".

Wed Dec 11 13:33:34 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c (main): Set obscure_syntax to tell re_search to
	allocate memory for the group registers.

Fri Dec  6 18:26:27 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tac.c, sort.c, csplit.c [POSIX]:  Use sigaction instead of
	signal, which POSIX doesn't have.
	* sort.c: Replace inthandler and huphandler with sighandler.
	* csplit.c (main): Only handle signals if they were not being
	ignored.

	* tr.c: POSIX_ME_HARDER -> POSIXLY_CORRECT.

Wed Dec  4 00:47:47 1991  Jim Meyering  (meyering at wombat)

	* tr.c (unquote): Reformat code so it doesn't go beyond column 80.
	* tr.c (squeeze_filter): Comment a little on why it's better
	to step through the input by two.
	* tr.c (set_initialize): Write a comment describing the function.
	* tr.c: Eliminated the variable `portability_warnings' and replaced
	references to it by references to `!posix_pedantic'.  One of the
	uses of portability_warnings had been wrong.

Tue Dec  3 14:03:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* tr.c: New program.

Sun Dec  1 15:07:35 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* linebuffer.[ch] (freebuffer): New function (used by cron).

Thu Oct 17 22:30:22 1991  David J. MacKenzie  (djm at wookumz.gnu.ai.mit.edu)

	* system.h, configure, Makefile.in: Don't include memory.h if
	STDC_HEADERS, removing need for MEMORY_H_MISSING.

Thu 17 Oct 1991 16:53:07  Jim Meyering  (meyering at wombat)

	* pr.c (print_page): REALLY fixed `extra newline at EOF' problem.
	Somehow, part of my patch didn't make it last time.

Sat Oct 12 12:04:47 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)

	* tail.c (pipe_lines, pipe_bytes): Initialize `first->next'.

	* cmp.c (cmp): Print EOF message to stderr, not stdout, for
	POSIX 1003.2.11.2.

	* sort.c (xfwrite): fwrite never returns < 0, so check if
	number written is number we asked to write.
	(fillbuf, main): fread never returns < 0, so check ferror instead.
	From Rainer Orth.

Tue Oct  8 18:07:08 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Really fixed `extra newline at EOF' problem.
	* (store_columns): Fixed bug that caused `pr -b -2' to coredump
	on files of certain lengths.

Fri Oct  4 22:30:25 1991  Jim Meyering  (meyering at churchy)

	* pr.c (print_page): Fixed to not add single spurious newline
	at EOF when using -t.

Wed Oct  2 01:02:05 1991  David J. MacKenzie  (djm at apple-gunkies)

	* pr.c (print_page): Don't pad the page if -t given.

	* csplit.c (load_buffer), sort.c (mergefps): Use bcopy, not memcpy.

	* memchr.c (memchr): Fix typo.

Thu Sep 26 12:35:00 1991  David J. MacKenzie  (djm at churchy.gnu.ai.mit.edu)


gnu/src/sum.c
/* sum -- checksum and count the blocks in a file
   Copyright (C) 1986, 1989, 1991 Free Software Foundation, Inc.

   This program is free software; you can redistribute it and/or modify
   it under the terms of the GNU General Public License as published by
   the Free Software Foundation; either version 2, or (at your option)
   any later version.

   This program is distributed in the hope that it will be useful,
   but WITHOUT ANY WARRANTY; without even the implied warranty of
   MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
   GNU General Public License for more details.

   You should have received a copy of the GNU General Public License
   along with this program; if not, write to the Free Software
   Foundation, Inc., 675 Mass Ave, Cambridge, MA 02139, USA.  */

/* Like BSD sum or SysV sum -r, except like SysV sum if -s option is given. */

/* Written by Kayvan Aghaiepour and David MacKenzie. */

#include <stdio.h>
#include <sys/types.h>
#include <getopt.h>
#include "system.h"
#ifdef OS2
#include <io.h>
#endif

int bsd_sum_file ();
int sysv_sum_file ();
void error ();

/* The name this program was run with. */
char *program_name;

/* Nonzero if any of the files read were the standard input. */
int have_read_stdin;

/* Right-rotate 32-bit integer variable C. */
#define ROTATE_RIGHT(c) if ((c) & 01) (c) = ((c) >>1) + 0x8000; else (c) >>= 1;

struct option longopts[] =
{
  {"sysv", 0, NULL, 's'},
  {NULL, 0, NULL, 0}
};

void
usage ()
{
  print_version();
  fprintf (stderr, "\
Usage: %s [-s] [--sysv] [file...]\n", program_name);
  exit (1);
}

void
main (argc, argv)
     int argc;
     char **argv;
{
  int errors = 0;
  int optc;
  int (*sum_func) () = bsd_sum_file;

  program_name = argv[0];
  have_read_stdin = 0;

  while ((optc = getopt_long (argc, argv, "s", longopts, (int *) 0)) != -1)
    {
      switch (optc)
	{
	case 's':
	  sum_func = sysv_sum_file;
	  break;

	case '?':
          usage();
	}
    }

  if (optind == argc)
    {
      if (isatty(0))
        usage();

      if ((*sum_func) ("-", 0) < 0)
	errors = 1;
    }
  else
    for (; optind < argc; optind++)
      if ((*sum_func) (argv[optind], 1) < 0)
	errors = 1;

  if (have_read_stdin && fclose (stdin) == EOF)
    error (1, errno, "-");
  exit (errors);
}

/* Calculate and print the rotated checksum and the size in 1K blocks
   of file FILE, or of the standard input if FILE is "-".
   If PRINT_NAME is nonzero, print FILE next to the checksum and size.
   The checksum varies depending on sizeof(int).
   Return 0 if successful, -1 if an error occurs. */

int
bsd_sum_file (file, print_name)
     char *file;
     int print_name;
{
  register FILE *fp;
  register unsigned long checksum = 0; /* The checksum mod 2^16. */
  register long total_bytes = 0; /* The number of bytes. */
  register int ch;		/* Each character read. */

  if (!strcmp (file, "-"))
    {
      fp = stdin;
      have_read_stdin = 1;
#ifdef OS2
      setmode(0, O_BINARY);
#endif
    }
  else
    {
#ifdef OS2
      fp = fopen (file, "rb");
#else
      fp = fopen (file, "r");
#endif
      if (fp == NULL)
	{
	  error (0, errno, "%s", file);
	  return -1;
	}
    }

  /* This algorithm seems to depend on sign extension in `ch' in order to
     give the right results.  Ick.  */
  while ((ch = getc (fp)) != EOF)
    {
      total_bytes++;
      ROTATE_RIGHT (checksum);
      checksum += ch;
      checksum &= 0xffff;	/* Keep it within bounds. */
    }

  if (ferror (fp))
    {
      error (0, errno, "%s", file);
      if (strcmp (file, "-"))
	fclose (fp);
      return -1;
    }

  if (strcmp (file, "-") && fclose (fp) == EOF)
    {
      error (0, errno, "%s", file);
      return -1;
    }

  printf ("%05lu %5ld", checksum, (total_bytes + 1024 - 1) / 1024);
  if (print_name)
    printf (" %s", file);
  putchar ('\n');

  return 0;
}

/* Calculate and print the checksum and the size in 512-byte blocks
   of file FILE, or of the standard input if FILE is "-".
   If PRINT_NAME is nonzero, print FILE next to the checksum and size.
   Return 0 if successful, -1 if an error occurs. */

int
sysv_sum_file (file, print_name)
     char *file;
     int print_name;
{
  int fd;
  unsigned char buf[8192];
  register int bytes_read;
  register unsigned long checksum = 0;
  long total_bytes = 0;

  if (!strcmp (file, "-"))
    {
      fd = 0;
      have_read_stdin = 1;
#ifdef OS2
      setmode(0, O_BINARY);
#endif
    }
  else
    {
#ifdef OS2
      fd = open (file, O_RDONLY | O_BINARY);
#else
      fd = open (file, O_RDONLY);
#endif
      if (fd == -1)
	{
	  error (0, errno, "%s", file);
	  return -1;
	}
    }

  while ((bytes_read = read (fd, buf, sizeof buf)) > 0)
    {
      register int i;

      for (i = 0; i < bytes_read; i++)
	checksum += buf[i];
      total_bytes += bytes_read;
    }

  if (bytes_read < 0)
    {
      error (0, errno, "%s", file);
      if (strcmp (file, "-"))
	close (fd);
      return -1;
    }

  if (strcmp (file, "-") && close (fd) == -1)
    {
      error (0, errno, "%s", file);
      return -1;
    }

  printf ("%lu %ld", checksum % 0xffff, (total_bytes + 512 - 1) / 512);
  if (print_name)
    printf (" %s", file);
  putchar ('\n');

  return 0;
}
